openmdao/components/implicit_func_comp.py

```diff
@@ -0,0 +1,191 @@
+
+
+class ImplicitFuncComp(ImplicitComponent):
+    """
+    An implicit component that wraps a python function.
+
+    Parameters
+    ----------
+    apply_nonlinear : function
+        The function to be wrapped by this Component.
+    solve_nonlinear : function or None
+        Optional function to perform a nonlinear solve.
+    linearize : function or None
+        Optional function to compute partial derivatives.
+    solve_linear : function or None
+        Optional function to perform a linear solve.
+    **kwargs : named args
+        Args passed down to ImplicitComponent.
+
+    Attributes
+    ----------
+    _apply_nonlinear_func : callable
+        The function wrapper used by this component.
+    _solve_nonlinear_func : function or None
+        Optional function to do a nonlinear solve.
+    solve_nonlinear : method
+        Local override of _solve_nonlinear method.
+    _solve_linear_func : function or None
+        Optional function to do a linear solve.
+    solve_linear : method
+        Local override of solve_linear method.
+    _linearize_func : function or None
+        Optional function to compute partial derivatives.
+    linearize : method
+        Local override of linearize method.
+    _linearize_info : object
+        Some state information to compute in _linearize_func and pass to _solve_linear_func
+    """
+
+    def __init__(self, apply_nonlinear, solve_nonlinear=None, linearize=None, solve_linear=None,
+                 **kwargs):
+        """
+        Initialize attributes.
+        """
+        super().__init__(**kwargs)
+        self._apply_nonlinear_func = omf.wrap(apply_nonlinear)
+        self._solve_nonlinear_func = solve_nonlinear
+        self._solve_linear_func = solve_linear
+        self._linearize_func = linearize
+        self._linearize_info = None
+        if solve_nonlinear:
+            self.solve_nonlinear = self._solve_nonlinear_
+        if linearize:
+            self.linearize = self._linearize_
+        if solve_linear:
+            self.solve_linear = self._solve_linear_
+
+    def setup(self):
+        """
+        Define out inputs and outputs.
+        """
+        optignore = {'is_option'}
+
+        for name, meta in self._apply_nonlinear_func.get_input_meta():
+            _check_var_name(self, name)
+            if 'is_option' in meta and meta['is_option']:
+                kwargs = _copy_with_ignore(meta, omf._allowed_declare_options_args,
+                                           ignore=optignore)
+                self.options.declare(name, **kwargs)
+            else:
+                kwargs = _copy_with_ignore(meta, omf._allowed_add_input_args)
+                self.add_input(name, **kwargs)
+
+        for i, (name, meta) in enumerate(self._apply_nonlinear_func.get_output_meta()):
+            if name is None:
+                raise RuntimeError(f"{self.msginfo}: Can't add output corresponding to return "
```

Comment:
Not clear this error has test coverage.

EDIT - found a few more; I guess this initial implementation doesn't yet have the tests for the error messages.